### ISSUE TITLE
Could not log in:  Grant successful but error accessing fc account:

### ISSUE THREAD

**Issue by dineshkumarsarangapani** [FlowCrypt/flowcrypt-browser#3302]
`Could not log in:  Grant successful but error accessing fc account: Error: error: -1 when POST-ing https://flowcrypt.com/api/account/login string: account,uuid,token -> (no standard err msg)`

I got this error and could not able to read messages from new user. Is there any other log I can fetch.?
I am using 

1. MAC - 10.14.6
2. Chrome - 87.0.4280.88 (Official Build) (x86_64)
3. FlowCrypt Plugin

**Comment by tomholub** [comment]
Hello - thank you for the report.

When you press F12 and then choose the `network` tab, you should see the network requests on the right hand side. One of them will fail, it would be interesting to know why it failed.

If it's not clear there, try the `console` tab instead, where it may also show an error.

**Comment by tomholub** [comment]
-1 typically means a network error, could be caused by actual network, or commonly a firewall or a some blocking extension like ublock, or even CORS issues.

**Comment by dineshkumarsarangapani** [comment]
@tomholub Thanks for the response. I see that it failed due to CORS issue.

`Failed to load resource: the server responded with a status of 401 (Unauthorized)
index.htm:1 Access to XMLHttpRequest at 'https://flowcrypt.com/blog/feed.xml' from origin 'chrome-extension://bnjglocicdkmhmoohhfkfkbbkejdhdgc' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
flowcrypt.com/blog/feed.xml:1 Failed to load resource: net::ERR_FAILED
index.htm:1 Access to XMLHttpRequest at 'https://flowcrypt.com/api/account/login' from origin 'chrome-extension://bnjglocicdkmhmoohhfkfkbbkejdhdgc' has been blocked by CORS policy: Request header field authorization is not allowed by Access-Control-Allow-Headers in preflight response.
jquery.min.js:2 POST https://flowcrypt.com/api/account/login net::ERR_FAILED
send @ jquery.min.js:2
ajax @ jquery.min.js:2
(anonymous) @ api.js:55
Api.ajax @ api.js:53
Api.apiCall @ api.js:182
FlowCryptComApi.request @ flowcrypt-com-api.js:99
FlowCryptComApi.loginWithOpenid @ flowcrypt-com-api.js:14
AccountServer.loginWithOpenid @ account-server.js:15
GoogleAuth.newAuthPopup @ google-auth.js:150
async function (async)
GoogleAuth.newAuthPopup @ google-auth.js:125
GoogleAuth.newOpenidAuthPopup @ google-auth.js:173
(anonymous) @ settings.js:384
index.htm:1 Access to XMLHttpRequest at 'https://flowcrypt.com/api/account/login' from origin 'chrome-extension://bnjglocicdkmhmoohhfkfkbbkejdhdgc' has been blocked by CORS policy: Request header field authorization is not allowed by Access-Control-Allow-Headers in preflight response.
flowcrypt.com/api/account/login:1 POST https://flowcrypt.com/api/account/login net::ERR_FAILED
index.htm:1 Access to XMLHttpRequest at 'https://flowcrypt.com/api/account/login' from origin 'chrome-extension://bnjglocicdkmhmoohhfkfkbbkejdhdgc' has been blocked by CORS policy: Request header field authorization is not allowed by Access-Control-Allow-Headers in preflight response.
`

I could also notice that the OPTIONS call was successful.

![image](https://user-images.githubusercontent.com/1185492/103985537-21047900-51af-11eb-9b19-0baece67b1e8.png)

**Comment by dineshkumarsarangapani** [comment]
Any solution to this?

**Comment by tomholub** [comment]
This is interesting - the error message you submitted seems correctly failing, and yet it does not seem to affect other users. If what you screenshotted is correct, I would expect this to be broken for everyone (which it's not). Maybe a newly enforced browser behavior (but only on some versions?). Very interesting.

@seisvelas could you have a look to understand how come it actually works for most people? If it's sending an authentication header, but the header is not allowed by CORS.. how come it works?

Either way we do have to fix this.

Command to debug with:

```
$ curl -i -XPOST https://flowcrypt.com/api/account/login --data '{}'
HTTP/1.1 400 Bad Request
Server: nginx
Date: Sun, 31 Jan 2021 16:27:13 GMT
Content-Type: application/json
Content-Length: 111
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Content-Type, api-version, user-account, user-uuid, user-token
Strict-Transport-Security: max-age=63072000; includeSubDomains; preload
```

**Comment by tomholub** [comment]
Of course the fix should land at https://github.com/FlowCrypt/flowcrypt-backend/ and should be quick / easy, I'll make another issue for that.

**Comment by seisvelas** [comment]
When I use the extension, my browser doesn't do the preflight check (`OPTIONS /api/account/login`) at all. 

`Access-Control-Allow-Headers` is supposed to work like this:

1. A request is initiated by a script
2. The browser checks whether the request is "simple"  (only uses standard headers, common HTTP method, etc)
3. If the request is not simple, do an OPTION request to see if the unusual method or nonstandard headers we're using are actually allowed

In this case, we make this request with nonstandard headers (for example, `api-version`), which should trigger a preflight check. The browser should look at the allowed headers in the preflight check response and notice that `Authorization` is not allowed (even though it's included in the request), and throw an error.

In our case, however, this does not happen because the browser is *usually* not sending the preflight check at all. Why not? Because `https://flowcrypt.com/*` is included in the "Site access" permissions of the FlowCrypt extension. 

![Screen Shot 2021-01-31 at 20 39 28](https://user-images.githubusercontent.com/44826516/106412218-6b6ec380-6404-11eb-8b0d-6087d9d0cb42.png)

If an extension has "Site access" permissions for a URL, then requests to that URL don't have to go through preflight check even if they are not simple. This is why FlowCrypt users are currently able to login.

Now the only mystery is why the browser of the user who reported this *is* doing the preflight check. Perhaps they removed `https://flowcrypt.com/*`'s "Site access" permission?

**Comment by seisvelas** [comment]
Presumably we want the extension's requests to pass preflight check were it to occur, in which case we'll likely have some other hurdles to overcome. For example, in the login request as made by the extension, credentials are set to 'include' (which we'd expect), yet we don't have a `Access-Control-Allow-Credentials` header to allow that. But even if we do set `Access-Control-Allow-Credentials: true`, we'll run into a new issue - If credentials mode is "include", then `Access-Control-Allow-Origin` cannot be `*`. So we would need to dynamically set the allowed origin to be the origin that made the request, instead of explicitly setting it to `*` as we have it now (which has the same effect in practice, it's just a hack to allow us to allow credentials).

So in order to make the extension's requests completely compliant, we need to add the authorization header to the allowed headers (https://github.com/FlowCrypt/flowcrypt-backend/pull/256), add `Access-Control-Allow-Credentials: true` to the response headers, and dynamically set the allowed origin to whatever origin is making the request.

(Although in practice the efficacy of doing all this is limited by the fact that the browser should ignore all of this anyway as long as the site access permissions are given flowcrypt.com in the FlowCrypt extension)

**Comment by dineshkumarsarangapani** [comment]
@seisvelas After enabling site access in my browser, I am not seeing this error again. 👍🏽 
Thanks for the support.

I don't know how the site access is turned off for that extension.

**Comment by seisvelas** [comment]
Glad to hear that did the trick, @dineshkumarsarangapani! 

@tomholub Continuing from my previous comment, this issue has exposed a few flaws with the way the extension and login API do CORS. Those issues are hidden in practice because the browser simply isn't doing preflight checks. But I'll make a separate issue for that, now that (I think) I understand how this should be done.

**Comment by tomholub** [comment]
Thank you! This explains why it indeed normally worked.
